src/olympia/versions/models.py

```diff
@@ -457,19 +457,34 @@ def from_upload(
             }
 
         compatible_apps = {}
-        for parsed_app in parsed_data.get('apps', []):
-            if parsed_app.appdata not in compatibility:
-                # If the user chose to explicitly deselect Firefox for Android
-                # we're not creating the respective `ApplicationsVersions`
-                # which will have this add-on then be listed only for
-                # Firefox specifically.
+        for avs_from_parsed_data in parsed_data.get('apps', []):
+            application = APP_IDS.get(avs_from_parsed_data.application)
+            if not application:
                 continue
-            avs = compatibility[parsed_app.appdata]
+            elif avs_from_parsed_data.locked_from_manifest:
+                # In that case, the manifest takes precedence.
+                avs = avs_from_parsed_data
```

Comment:
I can't do this as a suggestion because there's a deleted line in the middle, but 
```python
        for avs_from_parsed_data in parsed_data.get('apps', []):
            if avs_from_parsed_data.locked_from_manifest:
                # In that case, the manifest takes precedence.
                avs = avs_from_parsed_data
            application = APP_IDS.get(avs_from_parsed_data.application)
            if application not in compatibility:
                # In that case, the developer didn't include compatibility
                # information we have to follow, and they didn't select this
                # app at upload time, so we ignore that app.
                continue
            else:
```

Reply:
bc7178d6d999e9cd73e9422fcccc60e9f8b787b5